sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2ScanRelationPushDown.scala

```diff
@@ -30,6 +30,35 @@ object V2ScanRelationPushDown extends Rule[LogicalPlan] {
   import DataSourceV2Implicits._
 
   override def apply(plan: LogicalPlan): LogicalPlan = plan transformDown {
+    case RowLevelOperation(operation, cond, relation: DataSourceV2Relation, rewritePlan) =>
+      val mergeTable = relation.table.asMergeTable
+      val scanBuilder = mergeTable.newScanBuilder(relation.options)
+
+      val normalizedFilters = DataSourceStrategy.normalizeExprs(cond :: Nil, relation.output)
+      val (_, normalizedFiltersWithoutSubquery) =
+        normalizedFilters.partition(SubqueryExpression.hasSubquery)
+
+      val (pushedFilters, remainingFilters) = PushDownUtils.pushFilters(
```

Comment:
I have a question here for filter pushed down (since I am refactoring this `V2ScanRelationPushDown`). Are the `remainingFilters` going to be applied to V2Relation?

Reply:
Here I only care about filters that can be pushed into the source.

Do you have a PR for refactoring at hand, @huaxingao?

Reply:
I am refactoring `V2ScanRelationPushDown` in this PR https://github.com/apache/spark/pull/32049.
I have one more question for filter push down: in the doc of `RowLevelOperation`.`scanBuilder()`, it is said that `Some sources will avoid pushing filters into files (file granularity), while others will avoid pruning files within a partition (partition granularity)`. To avoid pushing down the filters that are mentioned above, do you need to change the current implementation of filter push down?

Reply:
Not really, we still want Spark to pass filters but it is up to the data source to apply them. For example, there may be implementations that apply filters to partitions but not files within partitions if they can atomically replace only partitions.

To sum up, I don't anticipate any major changes to the existing pushdown logic.